plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/building/BuildScopeAwareParallelLoaderProvider.java

```diff
@@ -0,0 +1,56 @@
+/**
+ * Copyright (c) 2017 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.ui.building;
+
+import org.eclipse.core.resources.IProject;
+import org.eclipse.emf.ecore.resource.ResourceSet;
+import org.eclipse.n4js.ts.scoping.builtin.BuiltInTypeScope;
+import org.eclipse.xtext.builder.resourceloader.IResourceLoader;
+import org.eclipse.xtext.builder.resourceloader.ResourceLoaderProviders.AbstractResourceLoaderProvider;
+import org.eclipse.xtext.resource.XtextResourceSet;
+import org.eclipse.xtext.ui.resource.IResourceSetProvider;
+
+/**
+ * Custom provider for parallel resource loader.
+ *
+ * Adjusted to return our own {@link BuildScopeAwareParallelLoader} and skip the registration of the
+ * {@link BuiltInTypeScope} in the loaders resource set.
+ */
+@SuppressWarnings("restriction")
+public class BuildScopeAwareParallelLoaderProvider extends AbstractResourceLoaderProvider {
+
+	private static final class ProviderOfEmptyResourceSet implements IResourceSetProvider {
+		@Override
+		public ResourceSet get(IProject project) {
+			return new XtextResourceSet();
+		}
+	}
+
+	private static final int nrOfThreads;
+
+	static {
+		int nProcessors = Runtime.getRuntime().availableProcessors();
+		int nThreads = Math.max(2, Math.min(4, nProcessors));
```

Comment:
@szarnekow  could you give (me) background for this estimate?
Why not use something like `Thread.activeCount()` ?

running following 
```java
	static {
		int nProcessors = Runtime.getRuntime().availableProcessors();
		int nThreads = Math.max(2, Math.min(4, nProcessors));
		nrOfThreads = nThreads;

		System.out.println("activeCount :: " + Thread.activeCount());
		System.out.println("availableProcessors :: " + nProcessors);
		System.out.println("nThreads :: " + nThreads);
	}
```
I get
```
activeCount :: 14
availableProcessors :: 8
nThreads :: 4
```

Reply:
so magic number based on experience. ok. (could be documented / commented).